### ISSUE TITLE
How to obtain local explanations that are interpretable as probabilities with EBM?

### ISSUE THREAD

**Issue by davidefiocco** [interpretml/interpret#595]
I am training EBMs as binary classifiers, and ideally I would like to extract at inference time each feature's contribution to the final predicted score for the positive class, **for a specific example**. Can this be done?

In other words, I would like the analog of what one can do with shap/shap using 

```python
# configure explainer so to compute additive probabilities
explainer = shap.TreeExplainer(classifier, model_output='probability', data = background_data)
shap_values = explainer.shap_values(test_df)
```

which allows to have the predicted value equal to the sum of the expected value plus explanations:

```python
idx = 42
classifier.predict_proba([test_df.iloc[idx]]) == explainer.expected_value + sum(shap_values[idx])
```

I had a look at https://github.com/interpretml/interpret/issues/157 and inspired by that I thought one could come up with something along the lines of

```python
import numpy as np

mean_score = 0.3  # 30% base probability on a reference background population
log_odds = np.array([1.0, -0.5])  # From EBMs, based on two features

mean_log_odds = np.log(mean_score / (1 - mean_score))
probs_with_features = 1 / (1 + np.exp(-(mean_log_odds + log_odds)))
    
# Calculate additive contributions... is this right though?
prob_contributions = probs_with_features - mean_score
```

But I am not sure at all that the above makes sense. Also, mind that I would need one single score per feature, and not having interaction terms.

**Comment by paulbkoch** [comment]
interpret.utils.inv_link(ebm.eval_terms(X).sum(axis=1) + ebm.intercept\_, ebm.link\_)

**Comment by paulbkoch** [comment]
Or, if you want to do it 100% yourself, we have the following code sample that can be run even without the interpret library:

https://interpret.ml/docs/ebm-internals-regression.html

**Comment by davidefiocco** [comment]
Hey @paulbkoch, thanks for your answer! I think I am... getting there but have a follow up question to clarify things further:

Say I am running:

```python
import numpy as np
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from interpret.glassbox import ExplainableBoostingClassifier
from interpret import show
import pandas as pd
from sklearn.metrics import accuracy_score

# Load breast cancer dataset
data = load_breast_cancer()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = data.target

# Split the data
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Initialize and train EBM model
ebm = ExplainableBoostingClassifier(
    inner_bags=10,
    outer_bags=5,
    learning_rate=0.01,
    interactions=1,
    random_state=42
)

# Train the model
ebm.fit(X_train, y_train)
```

I want now to run predictions, on a _specific_ record on my test set, and to do so I do:

```python
X_example = X_test.iloc[[0],:]
ebm.predict_proba(X_example)[0][1]
```

And I get that the score for the positive class is 0.971.

I would like now to "decompose" that 0.971 into the various (additive) probability contributions by the different features (e.g. `mean_radius`, `mean_texture`,...) and output a dictionary in which each key is the feature name and the value is the additive contribution.

Your suggestion

```
interpret.utils.inv_link(ebm.eval_terms(X_example).sum(axis=1) + ebm.intercept_, ebm.link_)
```

Allows me to obtain the class socres from `eval_terms`, am I getting this right?

However, ideally what I am looking for is the contribution of each feature to the score, for a specific example.

**Comment by Harsha-Nori** [comment]
Davide, good question. The main issue is that probabilities aren’t additive, unfortunately, because they are strictly bounded between 0 and 1. Most models like logistic regression, GAMs, EBMs, and Trees therefore choose to operate in logit space, which is what happens before the link function gets applied. You can use the eval terms function to see the relative contributions in additive logit space before it gets squished to a probability if you just remove the inv_link call above.

As an example, going from 0 to 1 logits adds a huge chunk of probability to the prediction (something like 20+%), but going from +5 +8 logits total adds much less than 1% (because probabilities are capped at 1, so you start going from 99.9 -> 99.99999 as you increase the logit towards infinity).

On my phone, hope this makes sense

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by paulbkoch** [comment]
Agree with everything @Harsha-Nori said.  I wanted to add that the values returned by eval_terms are also the local explanation values, in addition to being the values that you add to get a prediction in logit space.

**Comment by davidefiocco** [comment]
Thanks both @Harsha-Nori @paulbkoch. If I understand correctly, while eval terms are very useful for local explanations and also related to my problem, EBMs don't allow to generate additive, local explanations "à la SHAP". If that's the case, we can close this.

**Comment by paulbkoch** [comment]
@davidefiocco, SHAP has a link parameter which you can set to "logit", which is the analog to a classification EBM.  SHAP by default uses link="identity", which is fine in the case of SHAP since SHAP does not allow you to make predictions from the SHAP values, unlike with EBMs.  If you wanted a purely additive structure for probabilities, you would need to use the ExplainableBoostingRegressor, which uses the identity link.  You shouldn't do that however, since then you could get predicted probabilities that sum to more than 1, or less than 0.

There's a deep relationship between SHAP and EBMs in that the SHAP values from a mains-only EBM are exactly equal to the SHAP values, assuming you use the same link function for both.  This is a special property of EBMs, other GAMs, and linear models that is not shared with blackbox models such as traditional GBDTs.  SHAP has a dedicated explainer to handle EBMs for this reason: https://github.com/shap/shap/blob/master/shap/explainers/_additive.py

So, the outputs from eval_terms are actually three things simultaneously:
1) The local EBM explanations
2) The values that you add in order to make predictions with the EBM.
3) The exact SHAP values of the EBM (when using an equivalent link).

If you include interactions in an EBM, things become more complicated.  SHAP includes an option to generate pairwise SHAP values, and if you generated pairwise SHAP values on an EBM with interactions they would again maintain their perfect equivalence.

**Comment by davidefiocco** [comment]
Thanks @paulbkoch, that also clarifies.

For the time being I'll stick to something along the lines of

```python
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from lightgbm import LGBMClassifier
from sklearn.metrics import accuracy_score

# Load breast cancer dataset
data = load_breast_cancer()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = data.target

# Split the data
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Initialize the LightGBM Classifier
lgbm = LGBMClassifier(
    n_estimators=100,
    learning_rate=0.01,
    num_leaves=31,
    random_state=42
)

# Train the model
lgbm.fit(X_train, y_train)

# Predict and evaluate for a given example
X_example = X_test.iloc[[idx],:]
lgbm.predict_proba(X_example)[0][1]

# Generate explanation with the SHAP functionality of interpret
from interpret.blackbox import ShapKernel
from interpret import show

shap = ShapKernel(lgbm, X_train)
shap_local = shap.explain_local(X_example)
show(shap_local) 
```

As the last lines generate in the end exactly what I was looking for (using SHAP).
